src/main/java/com/enderio/base/common/handler/TravelHandler.java

```diff
@@ -97,27 +101,84 @@ public static boolean blockTeleport(Level level, Player player) {
     }
 
     public static Optional<Vec3> teleportPosition(Level level, Player player) {
-        Vec3 targetVec = player.position().add(0, player.getEyeHeight(), 0);
-        Vec3 lookVec = player.getLookAngle().normalize();
         @Nullable BlockPos target = null;
         double floorHeight = 0;
-        for (double i = BaseConfig.COMMON.ITEMS.TRAVELLING_BLINK_RANGE.get(); i >= 2; i -= 0.5) {
-            Vec3 v3d = targetVec.add(lookVec.scale(i));
-            target = new BlockPos((int) Math.round(v3d.x), (int) Math.round(v3d.y), (int) Math.round(v3d.z));
+
+        // inspired by Entity#pick
+        Vec3 playerPos = player.getEyePosition();
+        Vec3 lookVec = player.getLookAngle().normalize();
+        int range = BaseConfig.COMMON.ITEMS.TRAVELLING_BLINK_RANGE.get();
+        Vec3 toPos = playerPos.add(lookVec.scale(range));
+
+        ClipContext clipCtx = new ClipContext(playerPos, toPos, ClipContext.Block.OUTLINE, ClipContext.Fluid.NONE, null);
+        BlockHitResult bhr = level.clip(clipCtx);
+
+        // process the result
+        if (bhr.getType() == HitResult.Type.MISS) {
+            target = bhr.getBlockPos();
+        } else if (bhr.getType() == HitResult.Type.BLOCK) {
+            Direction dir = bhr.getDirection();
+            if (dir == Direction.UP) {
+                // warning: relies on the fact that isTeleportClear works with heights >= 1
```

Comment:
What does this mean? I'm struggling to understand from the comment alone what this refers to :)

Reply:
Yeah that comment was pretty unhelpful, I hope it's better now. At first that code chose the block above the target, but that was too high for carpets and too low for fences. Just choosing the target itself gets the correct height as long as isTeleportClear doesn't mind dealing with tall blocks.